### PR TITLE
Configure separate IAM roles for master & minion, create roles in docs

### PR DESCRIPTION
**Pull request by justinsb**

### CODE REVIEW

File: cluster/aws/util.sh [hunk 1]
```diff
@@ -237,8 +237,14 @@ function kube-up {
     ssh-keygen -f $AWS_SSH_KEY -N ''
   fi
 
-  aws iam get-instance-profile --instance-profile-name ${IAM_PROFILE} || {
-        echo "You need to set up an IAM profile and role for kubernetes"
+  detect-image
```
Comment (a-robinson) [inline cluster/aws/util.sh]: It looks like git won the battle on this file - detect-image isn't defined in this file in this branch, and you have some merge conflicts below.

File: docs/getting-started-guides/aws.md [hunk 2]
```diff
@@ -19,6 +19,86 @@
 2. ```cd kubernetes; make release```
 3. ```export PATH=$PATH:$PWD/_output/local/bin/<os>/<platform>```
 
+#### Create IAM instance profiles
+
+# TODO: Move to script?
```
Comment (a-robinson) [inline docs/getting-started-guides/aws.md]: Yeah, you don't have to do it now, but it'd be very nice if this was automated.

**Comment by justinsb** [comment]
I cleaned up my failed git merge - thanks.

I totally agree on the automation of the IAM roles.  I'm not sure if any other clouds have something similar which would help make this decision obvious.  One advantage of putting it into a script is that the JSON files could be files, and not heredocs (I guess I could make them files and still have commands-in-the-docs though).  The disadvantage is that it's less obvious to the user what's going on, but perhaps the user doesn't need to know at this getting-started stage.  I will ponder over pi-day!

**Comment by a-robinson** [comment]
Hmm, github still says this can't be merged. While you're at it, I noticed that the  "# TODO" lines in your getting-started guide markdown get converted to very large font in the generated page. If you already have to fix the commit to make it mergeable, you may want to remove the # signs from those TODOs as well.

I don't know if we've dealt with this on any other cloud - we certainly haven't in our GCE, GKE, and Vagrant providers. My inclination would be to put the actual details in the scripts and just explain in the guide that the script will create the two roles, but I haven't spent any time thinking it through. We should at least make the user aware of the types of resources that we're creating under their account, but we can only explain so much in a getting-started guide without it starting to look complicated.

**Comment by erictune** [comment]
Users are either going to read the scripts carefully before running them, or blindly trust them.  In either case, additional documentation about what it does won't be used.

**Comment by erictune** [comment]
I could see someone wanting to put a service account on their minion or master in GCE.   That can't be scripted.  So, seems okay to leave it unautomated here.

**Comment by justinsb** [comment]
Thanks for the excellent guidance.

I moved the IAM JSON config files out of the script to cluster/aws/templates/iam, and I now automatically create the instance roles if they do not exist already.

Getting started is still easy, therefore.  Power users can create their own instance roles and/or tweak the default permissions.

These permissions work for now, but I expect we will tweak the permissions over time.

File: cluster/aws/util.sh [hunk 3]
```diff
@@ -282,12 +282,45 @@ function add-tag {
...
+}
+
+# Creates the IAM roles (if they do not already exist)
+function ensure-iam-profiles {
+  aws iam get-instance-profile --instance-profile-name ${IAM_PROFILE_MASTER} || {
+    echo "Creating master IAM profile: ${IAM_PROFILE_MASTER}"
+    create-iam-profile ${IAM_PROFILE_MASTER}
+  }
+  aws iam get-instance-profile --instance-profile-name ${IAM_PROFILE_MINION} || {
+    echo "Creating master IAM profile: ${IAM_PROFILE_MINION}"
```
Comment (a-robinson) [inline cluster/aws/util.sh]: s/master/minion

**Comment by a-robinson** [comment]
LGTM other than the one comment. I'll merge once you fix that up.

**Comment by justinsb** [comment]
Thanks @a-robinson - you caught me copying and pasting!

**Pull request merged**